### PR TITLE
Support pattern variables declared under "is not"

### PR DESCRIPTION
**Pull request by gafter**
Implements https://github.com/dotnet/csharplang/issues/3369

LDM notes: https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-05-06.md#discussion

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests3.cs [hunk 1]
```diff
@@ -1984,36 +1985,30 @@ class Animal { }
 ";
             var compilation = CreateCompilation(source, parseOptions: TestOptions.RegularPreview);
             compilation.VerifyDiagnostics(
-                // (16,22): error CS8780: A variable may not be declared within a 'not' or 'or' pattern.
+                // (19,22): error CS8780: A variable may not be declared within a 'not' or 'or' pattern.
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests3.cs:1988]: >A variable may not be declared within a 'not' or 'or' pattern. [](start = 42, length = 62)

This message no longer seems appropriate

File: src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs [hunk 2]
```diff
@@ -1317,10 +1318,11 @@ void addSubpatternsForTuple(ImmutableArray<TypeWithAnnotations> elementTypes)
             TypeSymbol inputType,
             uint inputValEscape,
             bool hasErrors,
-            DiagnosticBag diagnostics)
+            DiagnosticBag diagnostics,
+            bool underIsPattern)
         {
-            const bool permitDesignations = false; // prevent designators under 'not'
-            var subPattern = BindPattern(node.Pattern, inputType, inputValEscape, permitDesignations, hasErrors, diagnostics);
+            bool permitDesignations = underIsPattern; // prevent designators under 'not' except under an is-pattern
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs:1324]: nit: consider removing local and just naming the arguments below.

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs [hunk 3]
```diff
@@ -959,7 +959,16 @@ public override BoundNode VisitIsPatternExpression(BoundIsPatternExpression node
         {
             Debug.Assert(!IsConditionalState);
             VisitRvalue(node.Expression);
-            VisitPattern(node.Pattern);
+
+            var pattern = node.Pattern;
+            bool negated = false;
+            while (pattern is BoundNegatedPattern n)
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:965]: >while (pattern is BoundNegatedPattern n) [](start = 12, length = 40)

might we encounter parentheses in this stack of `not`s? Or do parens patterns simply get omitted from bound nodes?

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests3.cs (hunk 1 above)
Comment (gafter) [reply src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests3.cs:1988]: It is appropriate, as there is only one specific syntactic circumstance in which we permit a pattern variable under a `not`. I don't think the diagnostic needs to explain that.

---
In reply to: [430768074](https://github.com/dotnet/roslyn/pull/44513#discussion_r430768074) [](ancestors = 430768074)

File: src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs (hunk 2 above)
Comment (gafter) [reply src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs:1324]: I did this for clarity.

---
In reply to: [430768444](https://github.com/dotnet/roslyn/pull/44513#discussion_r430768444) [](ancestors = 430768444)

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs (hunk 3 above)
Comment (gafter) [reply src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:965]: There is no bound pattern for the parens. They disappear in the bound nodes.

---
In reply to: [430768943](https://github.com/dotnet/roslyn/pull/44513#discussion_r430768943) [](ancestors = 430768943)

**Comment by gafter** [comment]
@jcouv I think I've responded to your comments. Do you have any others?

@dotnet/roslyn-compiler May I please have a second review?

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests3.cs [hunk 4]
```diff
@@ -5501,5 +5496,203 @@ static void Main()
                 );
             var compVerifier = CompileAndVerify(compilation, expectedOutput: expectedOutput);
         }
+
+        [Fact]
+        public void IsNot_01()
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests3.cs:5501]: >IsNot_01 [](start = 20, length = 8)

Consider adding a test that is not an `is`, but a switch expression arm. Or do we already have one?

Comment (gafter) [reply src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests3.cs:5501]: We already have many. There is no change of semantics for those.

---
In reply to: [430771480](https://github.com/dotnet/roslyn/pull/44513#discussion_r430771480) [](ancestors = 430771480)

**Comment by gafter** [comment]
@dotnet/roslyn-compiler May I please have a second review?

**Comment by gafter** [comment]
@dotnet/roslyn-compiler May I please have a second review?

**Comment by gafter** [comment]
@dotnet/roslyn-compiler May I please have a second review of this complete implementation of one of the C# 9.0 features? Would love to get it into the upcoming preview.

**Comment by gafter** [comment]
@dotnet/roslyn-compiler May I please have a second review?

**Pull request merged**